wasm/src/main/java/com/dylibso/chicory/wasm/io/InputStreamWasmInputStream.java

```diff
@@ -0,0 +1,139 @@
+package com.dylibso.chicory.wasm.io;
+
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+/**
+ * A WASM input stream based on {@link InputStream}.
+ */
+class InputStreamWasmInputStream extends WasmInputStream {
+
+    private long position;
+    private final BufferedInputStream is;
+    private final ByteBuffer littleBuffer =
+            ByteBuffer.wrap(new byte[8]).order(ByteOrder.LITTLE_ENDIAN);
+
+    InputStreamWasmInputStream(final InputStream is) {
+        this.is =
+                is instanceof BufferedInputStream
+                        ? (BufferedInputStream) is
+                        : new BufferedInputStream(is);
+    }
+
+    public long position() {
+        return position;
+    }
+
+    public void skip(long count) {
+        // todo: java 12+: is.skipNBytes(count);
+        if (count <= 0) {
+            return;
+        }
+
+        try {
+            long t;
+            while (count > 0) {
+                t = is.skip(count);
```

Comment:
Here seems like we are forcing a stricter semantics than the underlying `skip` mechanism, do you mind expanding on the reasons? Would it be possible to just reuse the underlying `is.skip` method?

Reply:
No, it is not possible because the contracts are in conflict. The use case for this method is that you want to unconditionally skip some amount of bytes (for example, skipping a section that we do not want to read). If fewer bytes are skipped, then the stream will be in an indeterminate state, which is not useful to the user. So if we do not provide the stricter contract, then the user of this method will have to do so themselves. The job of an API is to provide a useful abstraction over a lower-level construct, and this is one example of how we do so.

Note that Java 12 and later has a `skipNBytes` API on `InputStream` which implements this same logic for this same reason; once we can move beyond Java 11, we can replace most of this implementation code with a call to that method.

Reply:
Fair, thanks for explaining!